### ISSUE TITLE
[Relay] Where it is currently

### ISSUE THREAD

**Issue by cgyurgyik** [cucapra/futil#359]
I've never really laid out my thoughts on Relay and its direction, so this is an opportunity for me to do so before the demo on Tuesday. I intended on doing this last semester, but things were a bit shaky near the end for me.

This is a continuation of #276, and will discuss both next steps for VGG Net lowering, as well as improving the Relay frontend in general.

# Brittleness
There's lots of brittle stuff! My goal was to lower `VGG Net`; I definitely took some gross shortcuts. I'd rather be upfront and fix it now while the frontend is still relatively small in code size.

### 1. What the frontend currently supports
So the Relay IR that I've read such as `VGG Net` usually follow this form:
``` 
your_favorite_net(...) {
  let %x = foo(...);
  let %x1 = bar(x, ...);
  let %x2 = foobar(x1, ...);
}
// If you want to see the actual Relay IR for VGG Net:
//     python3 frontends/relay-futil/example.py vgg_net -r
```
So that's what I targeted. We first visit the function, which in turn visits the body:
https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/compiler.py#L90-L94
And then in [build_main_controls](https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/utilities.py#L80), we create the necessary connections to hook in each components wires, and a control flow, which is essentially just: `seq { foo; bar; foobar; }`.

There's functions for visiting all the Relay IR, [visit_var](https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/compiler.py#L55), [visit_let](https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/compiler.py#L62), etc. However, their current implementations only take into account that the IR will represent a Relay function call. I've also left some simply unsupported, such as `visit_constant` or visiting conditional branches, since they weren't required at the time.

### 2. AST (Where I need the most help to "get it right.")
In [relay-futil/futil_ast.py](https://github.com/cucapra/futil/blob/master/frontends/relay-futil/futil_ast.py), you'll see some AST support. Some of it is used, some of it not at all given that not all of Relay IR is supported. A lot of it seems extremely brittle or plain wrong, and needs a re-haul. For example, [FCell](https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/futil_ast.py#L145) just has a field for `relay_function`, which is a shortcut I took because "I'm an undergraduate and didn't know better", but I definitely knew better. It just worked at the time and I forgot to come back and think of a better way. Or, perhaps I was just missing the `invoke` piece at the time.

I *think* getting this right will make the rest of Relay -> FuTIL support much easier.

### 3. Supporting Tensors of many sizes.
Oh man. So a lot of these functions need support over many tensors sizes. Let's take a look at a
binary operation such as `add`.
For a 2-D tensor:
```
for (i = 0..N) 
  for (j = 0..M) 
    ret[i][j] := lhs[i][j] + rhs[i][j];
```
For a 3-D tensor,
```
for (i = 0..N) 
  for (j = 0..M) 
    for (k = 0..L) // Added this.
    ret[i][j][k] := lhs[i][j][k] + rhs[i][j][k]; // This changed.
```
If you look in [relay-futil/dahlia_lowering.py](https://github.com/cucapra/futil/blob/master/frontends/relay-futil/dahlia_lowering.py), you'll see I take advantage of this pattern to support endless array sizes. I basically assign the first iterator's name to the character `i`, and then if I need another iterator, i.e. there's another dimension, we take the next character, `j`. There's a function that does this: [PPDahliaLoop](https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/dahlia_lowering.py#L38). Then, we also need to change the number of index values in the body of the function, e.g. `x[i][j]` to `x[i][j][k]. An example of this can be seen in `nn.relu`:
https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/dahlia_lowering.py#L235-L240

We then take these indices, append them to the names of the variables to create our Dahlia program:
https://github.com/cucapra/futil/blob/d7f4b36b71efc3a5ba0b5ef5e1e74996a4a773cb/frontends/relay-futil/dahlia_lowering.py#L242-L244

This works for most of the functions that need to support multiple dimensions tensors, but it comes at the cost of difficult-to-read. There has to be a cleaner way to do this.

# VGG Net Lowering Steps
As discussed in #276, most everything has been completed thus far. There was a slight pause in tempo with the introduction of `invoke` and the end of the semester. The "last step" was implementing `nn.softmax`, which required an exponential operator. if you look in the code it was using an import of `std_exp`, but that's just incorrect right now.

Currently, a Relay function call is lowered from a Dahlia program to FuTIL component and externalized [here](https://github.com/cucapra/futil/blob/master/frontends/relay-futil/dahlia_lowering.py). 

With `invoke` finished, we should be able to instead just define a function at the Dahlia level, lower it, and invoke it in `main`. Instead of compiling separate Dahlia files for each component, we can now append all the function definitions in a single file, lower that, and then copy their component definitions to the final FuTIL file.

In an essence, this is the then-and-now: 
**Before**:
1. Write each function in separate Dahlia file. The next steps occur for each file.
2. Lower to FuTIL
3. Externalize
4. Copy component to final FuTIL file, and hook up wires to the respective component in `main`.

**After**:
1. Write function(s) in single Dahlia file.
2. Lower to FuTIL
3. Copy component to final FuTIL file, and `invoke` in `main`.

One thing to note, this will not be possible until (1) `invoke` lowering from Dahlia is finished, and (2) `invoke` supports memory as parameters to a function (for Dahlia at least).

# Testing
The test suite is pretty big since there are quite a few Relay function calls implemented, yet far from thorough. This means that bugs probably still exist, and there are many corner cases that aren't covered. I mean I'd love to write 3 or 4 tests per function, but that comes at much larger test suite.

For each Relay function I've implemented, I've added at least one test to make sure it works properly. For some of the less obvious functions such as `maxpool2d`, I've verified the outputs of an example or two against TensorFlow. I've also verified that running multiple Relay calls in a row, e.g. how `VGG Net` does it, works.
